torch/testing/_internal/common_methods_invocations.py

```diff
@@ -303,6 +338,32 @@ def sample_inputs(self, device, dtype, requires_grad=False, **kwargs):
             samples = self.sample_inputs_func(self, device, dtype, requires_grad)
         return samples
 
+    def sample_inputs(self, device, dtype, requires_grad=False, **kwargs):
+        """Returns a list of SampleInputCtxs.
+
+        The SampleInputCtxs are constructed by wrapping the result of
+        generate_sample_inputs.
+
+        If subclasses want to customize the generation of inputs, they should
+        override generate_sample_inputs(). On the contrary, end users should
+        always call sample_inputs().
+        """
+        samples = self.generate_sample_inputs(device, dtype, requires_grad, **kwargs)
+        # make sure that every SampleInput has a sensible name: if the name
+        # was not given, use the index instead
+        for i, sample in enumerate(samples):
```

Comment:
This is going to interfere with plans to make sample inputs use generators.

Why return the ctx and, if we take the context manager approach, not require tests just do:

```
with SampleGuard(sample):
```

for example? Seems like this complicates the process.

Reply:
> This is going to interfere with plans to make sample inputs use generators.

I'm not sure to understand how. We can still `enumerate()` even if `samples` is a generator. And if for some reason we really want `OpInfo.sample_inputs()` to return a generator instead of a list, we can just put a `yield` in the for loop

> Why return the ctx and, if we take the context manager approach, not require tests just do:
```
with SampleGuard(sample):
```
> for example? Seems like this complicates the process.

my idea is that by returning the context, users are *forced* to use the `with` statement, and so we are sure that exceptions will always be intercepted by the guard. In your proposal, what would happen if I forget the `with SampleGuard(sample)`? Can I still use `sample` even outside the `with`?

Reply:
Given that there are so few tests which use sample inputs I don't think we need to be very concerned with people forgetting to set the guard